dbms/src/Common/tests/gtest_fiber_rw_lock.cpp

```diff
@@ -0,0 +1,347 @@
+        typename Traits::VoidSharedFuture t1_entered_future(t1_entered.get_future());
+
+        typename Traits::Thread t1([&] {
+            auto future = t1_stop.get_future();
+            std::unique_lock lock(rw_lock);
+            t1_entered.set_value();
+            future.get();
+        });
+
+        std::atomic_int value = 0;
+        std::vector<int> read_values(5);
+        std::vector<typename Traits::Thread> read_threads;
+        auto read_f = [&](int idx) {
+            std::shared_lock lock(rw_lock);
+            read_values[idx] = value.load();
+        };
+        for (int i = 0; i < 5; ++i)
+            read_threads.emplace_back(read_f, i);
+
+        Traits::sleep();
+        value.store(1);
+
+        t1_stop.set_value();
+        t1.join();
+        for (auto & t : read_threads)
+            t.join();
+
+        for (const auto & v : read_values)
+            ASSERT_EQ(v, 1);
+    }
+
+    template <typename Traits>
+    void testReaderBlockWriter()
+    {
+        FiberRWLock rw_lock;
+        typename Traits::VoidPromise t1_stop;
+        typename Traits::VoidPromise t2_stopped;
+        typename Traits::VoidPromise t1_entered;
+        typename Traits::VoidSharedFuture t1_entered_future(t1_entered.get_future());
+
+        typename Traits::Thread t1([&] {
+            auto future = t1_stop.get_future();
+            std::shared_lock lock(rw_lock);
+            t1_entered.set_value();
+            future.get();
+        });
+
+        std::atomic_bool t2_entered = false;
+        typename Traits::Thread t2([&] {
+            t1_entered_future.get();
+            std::unique_lock lock(rw_lock);
+            t2_entered.store(true);
+            t2_stopped.set_value();
+        });
+
+        t1_entered_future.get();
+        ASSERT_EQ(t2_entered.load(), false);
+
+        t1_stop.set_value();
+        t2_stopped.get_future().get();
+        ASSERT_EQ(t2_entered.load(), true);
+
+        t1.join();
+        t2.join();
+    }
+
+    template <typename Traits>
+    void testWaitingWriterBlockReader()
+    {
+        FiberRWLock rw_lock;
+        typename Traits::VoidPromise t1_stop;
+        typename Traits::VoidPromise t1_entered;
+        typename Traits::VoidPromise t2_entered;
+        typename Traits::VoidPromise t3_stopped;
+        typename Traits::VoidSharedFuture t1_entered_future(t1_entered.get_future());
+        typename Traits::VoidSharedFuture t2_entered_future(t2_entered.get_future());
+
+        typename Traits::Thread t1([&] {
+            auto future = t1_stop.get_future();
+            std::shared_lock lock(rw_lock);
+            t1_entered.set_value();
+            future.get();
+        });
+
+        typename Traits::Thread t2([&] {
+            t1_entered_future.get();
+            std::unique_lock lock(rw_lock);
+            t2_entered.set_value();
+            
+        });
+
+        std::atomic_bool t3_entered = false;
+        typename Traits::Thread t3([&] {
```

Comment:
can we add a new test that makes t2 enter firstly, and then t3 enter, to ensure the assumption of write first?

Reply:
I tried but found it's hard in that we have to notify t3 after t2 starts waiting, but in this time t2 can't notify t3 (it's waiting!). Do you have any suggestion?